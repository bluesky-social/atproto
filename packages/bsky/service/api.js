'use strict' /* eslint-disable */

require('dd-trace') // Only works with commonjs
  .init({ logInjection: true })
  .tracer.use('express', {
    hooks: {
      request: (span, req) => {
        maintainXrpcResource(span, req)
      },
    },
  })

// Tracer code above must come before anything else
const path = require('path')
const assert = require('assert')
const { CloudfrontInvalidator } = require('@atproto/aws')
const {
  DatabaseCoordinator,
  PrimaryDatabase,
  ServerConfig,
  BskyAppView,
  ViewMaintainer,
  makeAlgos,
  PeriodicModerationActionReversal,
} = require('@atproto/bsky')

const main = async () => {
  const env = getEnv()
  assert(env.dbPrimaryPostgresUrl, 'missing configuration for db')
  const db = new DatabaseCoordinator({
    schema: env.dbPostgresSchema,
    primary: {
      url: env.dbPrimaryPostgresUrl,
      poolSize: env.dbPrimaryPoolSize || env.dbPoolSize,
      poolMaxUses: env.dbPoolMaxUses,
      poolIdleTimeoutMs: env.dbPoolIdleTimeoutMs,
    },
    replicas: env.dbReplicaPostgresUrls?.map((url, i) => {
      return {
        url,
        poolSize: env.dbPoolSize,
        poolMaxUses: env.dbPoolMaxUses,
        poolIdleTimeoutMs: env.dbPoolIdleTimeoutMs,
        tags: getTagsForIdx(env.dbReplicaTags, i),
      }
    }),
  })
  const cfg = ServerConfig.readEnv({
    port: env.port,
    version: env.version,
    dbPrimaryPostgresUrl: env.dbPrimaryPostgresUrl,
    dbReplicaPostgresUrls: env.dbReplicaPostgresUrls,
    dbReplicaTags: env.dbReplicaTags,
    dbPostgresSchema: env.dbPostgresSchema,
    publicUrl: env.publicUrl,
    didPlcUrl: env.didPlcUrl,
    imgUriSalt: env.imgUriSalt,
    imgUriKey: env.imgUriKey,
    imgUriEndpoint: env.imgUriEndpoint,
    blobCacheLocation: env.blobCacheLocation,
  })
  const cfInvalidator = env.cfDistributionId
    ? new CloudfrontInvalidator({
        distributionId: env.cfDistributionId,
        pathPrefix: cfg.imgUriEndpoint && new URL(cfg.imgUriEndpoint).pathname,
      })
    : undefined
  const algos = env.feedPublisherDid ? makeAlgos(env.feedPublisherDid) : {}
  const bsky = BskyAppView.create({
    db,
    config: cfg,
    imgInvalidator: cfInvalidator,
    algos,
  })
  // separate db needed for more permissions
  const migrateDb = new PrimaryDatabase({
    url: env.dbMigratePostgresUrl,
    schema: env.dbPostgresSchema,
    poolSize: 2,
  })
  const viewMaintainer = new ViewMaintainer(migrateDb)
  const viewMaintainerRunning = viewMaintainer.run()

  const periodicModerationActionReversal = new PeriodicModerationActionReversal(
    bsky.ctx,
  )
  const periodicModerationActionReversalRunning =
    periodicModerationActionReversal.run()

  await bsky.start()
  // Graceful shutdown (see also https://aws.amazon.com/blogs/containers/graceful-shutdowns-with-ecs/)
  process.on('SIGTERM', async () => {
    // Gracefully shutdown periodic-moderation-action-reversal before destroying bsky instance
    periodicModerationActionReversal.destroy()
    await periodicModerationActionReversalRunning
    await bsky.destroy()
    viewMaintainer.destroy()
    await viewMaintainerRunning
    await migrateDb.close()
  })
}

const getEnv = () => ({
  port: parseInt(process.env.PORT),
  version: process.env.BSKY_VERSION,
  dbMigratePostgresUrl:
    process.env.DB_MIGRATE_POSTGRES_URL || process.env.DB_PRIMARY_POSTGRES_URL,
  dbPrimaryPostgresUrl: process.env.DB_PRIMARY_POSTGRES_URL,
  dbPrimaryPoolSize: maybeParseInt(process.env.DB_PRIMARY_POOL_SIZE),
  dbReplicaPostgresUrls: process.env.DB_REPLICA_POSTGRES_URLS
    ? process.env.DB_REPLICA_POSTGRES_URLS.split(',')
    : undefined,
  dbReplicaTags: {
    '*': getTagIdxs(process.env.DB_REPLICA_TAGS_ANY), // e.g. DB_REPLICA_TAGS_ANY=0,1
    timeline: getTagIdxs(process.env.DB_REPLICA_TAGS_TIMELINE),
    feed: getTagIdxs(process.env.DB_REPLICA_TAGS_FEED),
    search: getTagIdxs(process.env.DB_REPLICA_TAGS_SEARCH),
    thread: getTagIdxs(process.env.DB_REPLICA_TAGS_THREAD),
  },
  dbPostgresSchema: process.env.DB_POSTGRES_SCHEMA || undefined,
  dbPoolSize: maybeParseInt(process.env.DB_POOL_SIZE),
  dbPoolMaxUses: maybeParseInt(process.env.DB_POOL_MAX_USES),
  dbPoolIdleTimeoutMs: maybeParseInt(process.env.DB_POOL_IDLE_TIMEOUT_MS),
  publicUrl: process.env.PUBLIC_URL,
  didPlcUrl: process.env.DID_PLC_URL,
  imgUriSalt: process.env.IMG_URI_SALT,
  imgUriKey: process.env.IMG_URI_KEY,
  imgUriEndpoint: process.env.IMG_URI_ENDPOINT,
  blobCacheLocation: process.env.BLOB_CACHE_LOC,
  cfDistributionId: process.env.CF_DISTRIBUTION_ID,
  feedPublisherDid: process.env.FEED_PUBLISHER_DID,
})

/**
 * @param {Record<string, number[]>} tags
 * @param {number} idx
 */
const getTagsForIdx = (tagMap, idx) => {
  const tags = []
  for (const [tag, indexes] of Object.entries(tagMap)) {
    if (indexes.includes(idx)) {
      tags.push(tag)
    }
  }
  return tags
}

/**
 * @param {string} str
 */
const getTagIdxs = (str) => {
  return str ? str.split(',').map((item) => parseInt(item, 10)) : []
}

const maybeParseInt = (str) => {
  const parsed = parseInt(str)
  return isNaN(parsed) ? undefined : parsed
}

const maintainXrpcResource = (span, req) => {
  // Show actual xrpc method as resource rather than the route pattern
  if (span && req.originalUrl?.startsWith('/xrpc/')) {
    span.setTag(
      'resource.name',
      [
        req.method,
        path.posix.join(req.baseUrl || '', req.path || '', '/').slice(0, -1), // Ensures no trailing slash
      ]
        .filter(Boolean)
        .join(' '),
    )
  }
}

main()
