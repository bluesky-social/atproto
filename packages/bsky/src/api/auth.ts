import express from 'express'
import * as uint8arrays from 'uint8arrays'
import { AuthRequiredError } from '@atproto/xrpc-server'

const BASIC = 'Basic '
const BEARER = 'Bearer '

// @TODO(bsky) treating did as a bearer, just a placeholder for now.
export const authVerifier = (ctx: {
  req: express.Request
  res: express.Response
}) => {
  const { authorization = '' } = ctx.req.headers
  if (!authorization.startsWith(BEARER)) {
    throw new AuthRequiredError()
  }
  const did = authorization.replace(BEARER, '').trim()
  if (!did.startsWith('did:')) {
    throw new AuthRequiredError()
  }
  return { credentials: { did } }
}

export const authOptionalVerifier = (ctx: {
  req: express.Request
  res: express.Response
}) => {
  if (!ctx.req.headers.authorization) {
    return { credentials: { did: null } }
  }
  return authVerifier(ctx)
}

export const adminVerifier =
  (adminPassword: string) =>
  (ctx: { req: express.Request; res: express.Response }) => {
    const { authorization = '' } = ctx.req.headers
    const parsed = parseBasicAuth(authorization)
    if (!parsed) {
      throw new AuthRequiredError()
    }
    const { username, password } = parsed
    if (username !== 'admin' || password !== adminPassword) {
      throw new AuthRequiredError()
    }
    return { credentials: { admin: true } }
  }

export const parseBasicAuth = (
  token: string,
): { username: string; password: string } | null => {
  if (!token.startsWith(BASIC)) return null
  const b64 = token.slice(BASIC.length)
  let parsed: string[]
  try {
    parsed = uint8arrays
      .toString(uint8arrays.fromString(b64, 'base64pad'), 'utf8')
      .split(':')
  } catch (err) {
    return null
  }
  const [username, password] = parsed
  if (!username || !password) return null
  return { username, password }
}
