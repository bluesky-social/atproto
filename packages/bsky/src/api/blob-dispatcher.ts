import { isUnicastIp, unicastLookup } from '@atproto-labs/fetch-node'
import { Agent, Dispatcher, Pool, RetryAgent } from 'undici'

import { ServerConfig } from '../config'
import { RETRYABLE_HTTP_STATUS_CODES } from '../util/retry'

export function createBlobDispatcher(cfg: ServerConfig): Dispatcher {
  const baseDispatcher = new Agent({
    allowH2: cfg.proxyAllowHTTP2, // This is experimental
    headersTimeout: cfg.proxyHeadersTimeout,
    maxResponseSize: cfg.proxyMaxResponseSize,
    bodyTimeout: cfg.proxyBodyTimeout,
    factory: cfg.disableSsrfProtection
      ? undefined
      : (origin, opts) => {
          const { protocol, hostname } =
            origin instanceof URL ? origin : new URL(origin)
          if (protocol !== 'https:') {
            throw new Error(`Forbidden protocol "${protocol}"`)
          }
          if (isUnicastIp(hostname) === false) {
            throw new Error('Hostname resolved to non-unicast address')
          }
          return new Pool(origin, opts)
        },
    connect: {
      lookup: cfg.disableSsrfProtection ? undefined : unicastLookup,
    },
  })

  return cfg.proxyMaxRetries > 0
    ? new RetryAgent(baseDispatcher, {
        statusCodes: [...RETRYABLE_HTTP_STATUS_CODES],
        methods: ['GET', 'HEAD'],
        maxRetries: cfg.proxyMaxRetries,
      })
    : baseDispatcher
}
