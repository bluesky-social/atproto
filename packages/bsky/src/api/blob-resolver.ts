import { pipeline, Readable } from 'stream'
import express from 'express'
import createError from 'http-errors'
import axios, { AxiosError } from 'axios'
import { CID } from 'multiformats/cid'
import { ensureValidDid } from '@atproto/identifier'
import { VerifyCidTransform } from '@atproto/common'
import { NoResolveDidError } from '@atproto/did-resolver'
import AppContext from '../context'
import { httpLogger as log } from '../logger'
import { retryHttp } from '../util/retry'

// Resolve and verify blob from its origin host

export const createRouter = (ctx: AppContext): express.Router => {
  const router = express.Router()

  router.get('/blob/:did/:cid', async function (req, res, next) {
    try {
      const { did, cid: cidStr } = req.params
      try {
        ensureValidDid(did)
      } catch (err) {
        return next(createError(400, 'Invalid did'))
      }
      let cid: CID
      try {
        cid = CID.parse(cidStr)
      } catch (err) {
        return next(createError(400, 'Invalid cid'))
      }

      const { pds } = await ctx.didResolver.resolveAtpData(did) // @TODO cache did info
      const blobResult = await retryHttp(() =>
        getBlob({ pds, did, cid: cidStr }),
      )

      const imageStream: Readable = blobResult.data
      const verifyCid = new VerifyCidTransform(cid)

      // Send chunked response, destroying stream early (before
      // closing chunk) if the bytes don't match the expected cid.
      res.statusCode = 200
      res.setHeader(
        'content-type',
        blobResult.headers['content-type'] || 'application/octet-stream',
      )
      pipeline(imageStream, verifyCid, res, (err) => {
        if (err) {
          log.warn(
            { err, did, cid: cidStr, pds },
            'blob resolution failed during transmission',
          )
        }
      })
    } catch (err) {
      if (err instanceof AxiosError) {
        if (err.code === AxiosError.ETIMEDOUT) {
          log.warn(
            { host: err.request?.host, path: err.request?.path },
            'blob resolution timeout',
          )
          return next(createError(504)) // Gateway timeout
        }
        if (!err.response || err.response.status >= 500) {
          log.warn(
            { host: err.request?.host, path: err.request?.path },
            'blob resolution failed upstream',
          )
          return next(createError(502))
        }
        return next(createError(404, 'Blob not found'))
      }
      if (err instanceof NoResolveDidError) {
        return next(createError(404, 'Blob not found'))
      }
      return next(err)
    }
  })

  return router
}

async function getBlob(opts: { pds: string; did: string; cid: string }) {
  const { pds, did, cid } = opts
  return axios.get(`${pds}/xrpc/com.atproto.sync.getBlob`, {
    params: { did, cid },
    decompress: true,
    responseType: 'stream',
    timeout: 2000, // 2sec of inactivity on the connection
  })
}
