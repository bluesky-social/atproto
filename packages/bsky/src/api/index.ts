import { Server } from '../lexicon'
import AppContext from '../context'
import getTimeline from './app/bsky/feed/getTimeline'
import getAuthorFeed from './app/bsky/feed/getAuthorFeed'
import getLikes from './app/bsky/feed/getLikes'
import getPostThread from './app/bsky/feed/getPostThread'
import getProfile from './app/bsky/actor/getProfile'
import getProfiles from './app/bsky/actor/getProfiles'
import getRepostedBy from './app/bsky/feed/getRepostedBy'
import getFollowers from './app/bsky/graph/getFollowers'
import getFollows from './app/bsky/graph/getFollows'
import searchActors from './app/bsky/actor/searchActors'
import searchActorsTypeahead from './app/bsky/actor/searchActorsTypeahead'
import getSuggestions from './app/bsky/actor/getSuggestions'
import unspecced from './app/bsky/unspecced'

export * as health from './health'

export * as blobResolver from './blob-resolver'

export default function (server: Server, ctx: AppContext) {
  getTimeline(server, ctx)
  getAuthorFeed(server, ctx)
  getLikes(server, ctx)
  getPostThread(server, ctx)
  getProfile(server, ctx)
  getProfiles(server, ctx)
  getRepostedBy(server, ctx)
  getFollowers(server, ctx)
  getFollows(server, ctx)
  searchActors(server, ctx)
  searchActorsTypeahead(server, ctx)
  getSuggestions(server, ctx)
  unspecced(server, ctx)
  return server
}
