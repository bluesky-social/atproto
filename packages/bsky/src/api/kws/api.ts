import express, { RequestHandler } from 'express'
import { httpLogger as log } from '../../logger'
import {
  AppContextWithKwsClient,
  KwsVerificationIntermediateQuery,
  KwsVerificationQuery,
  verificationIntermediateQuerySchema,
} from './types'
import {
  createStashEvent,
  getClientUa,
  parseExternalPayload,
  parseStatus,
  validateSignature,
} from './util'

const validateRequest = (
  ctx: AppContextWithKwsClient,
  req: express.Request,
): KwsVerificationQuery => {
  try {
    const intermediate: KwsVerificationIntermediateQuery =
      verificationIntermediateQuerySchema.parse({
        externalPayload: req.query.externalPayload,
        signature: req.query.signature,
        status: req.query.status,
      })

    const data = `${intermediate.status}:${intermediate.externalPayload}`
    validateSignature(
      ctx.cfg.kws.verificationSecret,
      data,
      intermediate.signature,
    )

    return {
      ...intermediate,
      externalPayload: parseExternalPayload(intermediate.externalPayload),
      status: parseStatus(intermediate.status),
    }
  } catch (err) {
    throw new Error('Invalid KWS API request', { cause: err })
  }
}

export const verificationHandler =
  (ctx: AppContextWithKwsClient): RequestHandler =>
  async (req: express.Request, res: express.Response) => {
    let actorDid: string | undefined
    try {
      const query = validateRequest(ctx, req)
      const {
        externalPayload,
        status: { verified },
      } = query
      if (!verified) {
        throw new Error(
          'Unexpected KWS verification response call with unverified status',
        )
      }

      const { actorDid: externalPayloadActorDid, attemptId } = externalPayload
      actorDid = externalPayloadActorDid
      // Assumes `app.set('trust proxy', ...)` configured with `true` or specific values.
      const completeIp = req.ip
      const completeUa = getClientUa(req)
      await createStashEvent(ctx, {
        actorDid,
        attemptId,
        completeIp,
        completeUa,
        status: 'assured',
      })
      return res
        .status(302)
        .setHeader(
          'Location',
          `${ctx.cfg.kws.redirectUrl}?${new URLSearchParams({ actorDid, result: 'success' })}`,
        )
        .end()
    } catch (err) {
      log.error({ err }, 'Failed to handle KWS verification response')

      return res
        .status(302)
        .setHeader(
          'Location',
          `${ctx.cfg.kws.redirectUrl}?${new URLSearchParams({ ...(actorDid ? { actorDid } : {}), result: 'unknown' })}`,
        )
        .end()
    }
  }
