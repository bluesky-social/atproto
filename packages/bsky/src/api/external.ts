import { Router } from 'express'
import { AppContext } from '../context'
import * as kwsApi from './kws'

export const createRouter = (ctx: AppContext): Router => {
  const router = Router()

  if (ctx.kwsClient) {
    router.use('/kws', kwsApi.createRouter(ctx))
  }

  return router
}
