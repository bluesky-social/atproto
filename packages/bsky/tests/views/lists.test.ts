import { AtpAgent } from '@atproto/api'
import { TestNetwork, SeedClient, basicSeed } from '@atproto/dev-env'
import { forSnapshot } from '../_util'
import { ids } from '../../src/lexicon/lexicons'

describe('bsky actor likes feed views', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient

  let curateList: string
  let referenceList: string
  let alice: string
  let eve: string
  let frankie: string
  let greta: string

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_actor_lists',
    })
    agent = network.bsky.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await sc.createAccount('eve', {
      handle: 'eve.test',
      email: 'eve@eve.com',
      password: 'hunter2',
    })
    await sc.createAccount('frankie', {
      handle: 'frankie.test',
      email: 'frankie@frankie.com',
      password: '2hunter2real',
    })
    await sc.createAccount('greta', {
      handle: 'greta.test',
      email: 'greta@greta.com',
      password: 'hunter4real',
    })

    const newRefList = await sc.createList(
      sc.dids.eve,
      'blah starter pack list!',
      'reference',
    )
    const newCurrList = await sc.createList(
      sc.dids.eve,
      'blah curate list!',
      'curate',
    )

    await sc.addToList(sc.dids.eve, sc.dids.eve, newRefList)
    await sc.addToList(sc.dids.eve, sc.dids.bob, newRefList)
    await sc.addToList(sc.dids.eve, sc.dids.frankie, newRefList)

    await sc.addToList(sc.dids.eve, sc.dids.eve, newCurrList)
    await sc.addToList(sc.dids.eve, sc.dids.bob, newCurrList)
    await sc.addToList(sc.dids.eve, sc.dids.frankie, newCurrList)

    await sc.block(sc.dids.frankie, sc.dids.eve)

    await network.processAll()
    curateList = newCurrList.uriStr
    referenceList = newRefList.uriStr
    alice = sc.dids.alice
    eve = sc.dids.eve
    frankie = sc.dids.frankie
    greta = sc.dids.greta
  })

  afterAll(async () => {
    await network.close()
  })

  it('does not include reference lists in getActorLists', async () => {
    await sc.createList(eve, 'cool curate list', 'curate')
    await network.processAll()
    const view = await agent.api.app.bsky.graph.getLists({
      actor: eve,
    })
    expect(view.data.lists.length).toBe(2)
    expect(forSnapshot(view.data.lists)).toMatchSnapshot()
  })

  it('supports using a handle as getList actor param', async () => {
    const view = await agent.app.bsky.graph.getLists({
      actor: 'eve.test',
    })
    expect(view.data.lists.length).toBe(2)
    expect(forSnapshot(view.data.lists)).toMatchSnapshot()
  })

  it('does not include users with creator block relationship in reference lists for non-creator, in-list viewers', async () => {
    const curView = await agent.api.app.bsky.graph.getList(
      {
        list: curateList,
      },
      {
        headers: await network.serviceHeaders(frankie, ids.AppBskyGraphGetList),
      },
    )
    expect(curView.data.items.length).toBe(2)
    expect(forSnapshot(curView.data.items)).toMatchSnapshot()

    const refView = await agent.api.app.bsky.graph.getList(
      { list: referenceList },
      {
        headers: await network.serviceHeaders(frankie, ids.AppBskyGraphGetList),
      },
    )
    expect(refView.data.items.length).toBe(2)
    expect(forSnapshot(refView.data.items)).toMatchSnapshot()
  })

  it('does not include users with creator block relationship in reference lists for non-creator, not-in-list viewers', async () => {
    const curView = await agent.api.app.bsky.graph.getList(
      {
        list: curateList,
      },
      { headers: await network.serviceHeaders(greta, ids.AppBskyGraphGetList) },
    )
    expect(curView.data.items.length).toBe(2)
    expect(forSnapshot(curView.data.items)).toMatchSnapshot()

    const refView = await agent.api.app.bsky.graph.getList(
      { list: referenceList },
      { headers: await network.serviceHeaders(greta, ids.AppBskyGraphGetList) },
    )
    expect(refView.data.items.length).toBe(2)
    expect(forSnapshot(refView.data.items)).toMatchSnapshot()
  })

  it('does not include users with creator block relationship in reference and curate lists for signed-out viewers', async () => {
    const curView = await agent.api.app.bsky.graph.getList({
      list: curateList,
    })
    expect(curView.data.items.length).toBe(2)
    expect(forSnapshot(curView.data.items)).toMatchSnapshot()

    const refView = await agent.api.app.bsky.graph.getList({
      list: referenceList,
    })
    expect(refView.data.items.length).toBe(2)
    expect(forSnapshot(refView.data.items)).toMatchSnapshot()
  })

  it('does include users with creator block relationship in reference lists for creator', async () => {
    const curView = await agent.api.app.bsky.graph.getList(
      { list: curateList },
      { headers: await network.serviceHeaders(eve, ids.AppBskyGraphGetList) },
    )
    expect(curView.data.items.length).toBe(3)
    expect(forSnapshot(curView.data.items)).toMatchSnapshot()

    const refView = await agent.api.app.bsky.graph.getList(
      { list: referenceList },
      { headers: await network.serviceHeaders(eve, ids.AppBskyGraphGetList) },
    )
    expect(refView.data.items.length).toBe(3)
    expect(forSnapshot(refView.data.items)).toMatchSnapshot()
  })

  it('does return all users regardless of creator block relationship in moderation lists', async () => {
    const blockList = await sc.createList(eve, 'block list', 'mod')
    await sc.addToList(eve, frankie, blockList)
    await sc.addToList(eve, greta, blockList)
    await sc.block(frankie, greta)
    await network.processAll()

    const view = await agent.api.app.bsky.graph.getList(
      { list: blockList.uriStr },
      { headers: await network.serviceHeaders(alice, ids.AppBskyGraphGetList) },
    )
    expect(view.data.items.length).toBe(2)
    expect(forSnapshot(view.data.items)).toMatchSnapshot()
  })
})
