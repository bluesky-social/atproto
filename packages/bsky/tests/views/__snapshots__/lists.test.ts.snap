// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bsky actor likes feed views does include users with creator block relationship in reference lists for creator 1`] = `
Array [
  Object {
    "subject": Object {
      "did": "user(0)",
      "handle": "frankie.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": true,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(2)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(1)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
  Object {
    "subject": Object {
      "did": "user(3)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(2)",
  },
]
`;

exports[`bsky actor likes feed views does include users with creator block relationship in reference lists for creator 2`] = `
Array [
  Object {
    "subject": Object {
      "did": "user(0)",
      "handle": "frankie.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": true,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(2)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(1)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
  Object {
    "subject": Object {
      "did": "user(3)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(2)",
  },
]
`;

exports[`bsky actor likes feed views does not include reference lists in getActorLists 1`] = `
Array [
  Object {
    "cid": "cids(0)",
    "creator": Object {
      "did": "user(0)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "listItemCount": 0,
    "name": "cool curate list",
    "purpose": "app.bsky.graph.defs#curatelist",
    "uri": "record(0)",
  },
  Object {
    "cid": "cids(1)",
    "creator": Object {
      "did": "user(0)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "listItemCount": 3,
    "name": "blah curate list!",
    "purpose": "app.bsky.graph.defs#curatelist",
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference and curate lists for signed-out viewers 1`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference and curate lists for signed-out viewers 2`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference lists for non-creator, in-list viewers 1`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "blocking": "record(2)",
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference lists for non-creator, in-list viewers 2`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "blocking": "record(2)",
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference lists for non-creator, not-in-list viewers 1`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does not include users with creator block relationship in reference lists for non-creator, not-in-list viewers 2`] = `
Array [
  Object {
    "subject": Object {
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(0)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "description": "hi im bob label_me",
      "did": "user(0)",
      "displayName": "bobby",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(2)",
      "handle": "eve.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views does return all users regardless of creator block relationship in moderation lists 1`] = `
Array [
  Object {
    "subject": Object {
      "did": "user(0)",
      "handle": "greta.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(0)",
  },
  Object {
    "subject": Object {
      "did": "user(1)",
      "handle": "frankie.test",
      "labels": Array [],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "uri": "record(1)",
  },
]
`;

exports[`bsky actor likes feed views supports using a handle as getList actor param 1`] = `
Array [
  Object {
    "cid": "cids(0)",
    "creator": Object {
      "did": "user(0)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "listItemCount": 0,
    "name": "cool curate list",
    "purpose": "app.bsky.graph.defs#curatelist",
    "uri": "record(0)",
  },
  Object {
    "cid": "cids(1)",
    "creator": Object {
      "did": "user(0)",
      "handle": "eve.test",
      "labels": Array [],
    },
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "listItemCount": 3,
    "name": "blah curate list!",
    "purpose": "app.bsky.graph.defs#curatelist",
    "uri": "record(1)",
  },
]
`;
